Signal/src/util/ThreadUtil.m

```diff
@@ -392,87 +435,65 @@ + (ThreadDynamicInteractions *)ensureDynamicInteractionsForThread:(TSThread *)th
             }
         }
 
+        BOOL shouldHaveContactOffers
+            = (shouldHaveBlockOffer || shouldHaveAddToContactsOffer || shouldHaveAddToProfileWhitelistOffer);
+        if (isContactThread) {
+            TSContactThread *contactThread = (TSContactThread *)thread;
+            // TODO: Set this property.
+            if (contactThread.hasDismissedOffers) {
+                shouldHaveContactOffers = NO;
+            }
+        }
+
         // We use these offset to control the ordering of the offers and indicators.
-        const int kAddToProfileWhitelistOfferOffset = -4;
-        const int kBlockOfferOffset = -3;
-        const int kAddToContactsOfferOffset = -2;
-        const int kUnreadIndicatorOfferOffset = -1;
+        const int kUnreadIndicatorOffset = -1;
 
         // We want the offers to be the first interactions in their
         // conversation's timeline, so we back-date them to slightly before
         // the first message - or at an aribtrary old timestamp if the
         // conversation has no messages.
-        long long startOfConversationTimestamp
-            = (long long)(firstMessage ? firstMessage.timestampForSorting : 1000);
-
-        if (existingBlockOffer && !shouldHaveBlockOffer) {
-            DDLogInfo(@"%@ Removing block offer: %@ (%llu)",
-                self.tag,
-                existingBlockOffer.uniqueId,
-                existingBlockOffer.timestampForSorting);
-            [existingBlockOffer removeWithTransaction:transaction];
-        } else if (!existingBlockOffer && shouldHaveBlockOffer) {
-            DDLogInfo(@"Creating block offer for unknown contact");
-
-            uint64_t blockOfferTimestamp = (uint64_t)(startOfConversationTimestamp + kBlockOfferOffset);
-            NSString *recipientId = ((TSContactThread *)thread).contactIdentifier;
-
-            TSMessage *offerMessage =
-                [OWSUnknownContactBlockOfferMessage unknownContactBlockOfferMessage:blockOfferTimestamp
-                                                                             thread:thread
-                                                                          contactId:recipientId];
-            [offerMessage saveWithTransaction:transaction];
-
-            DDLogInfo(@"%@ Creating block offer: %@ (%llu)",
-                self.tag,
-                offerMessage.uniqueId,
-                offerMessage.timestampForSorting);
+        uint64_t contactOffersTimestamp = [NSDate ows_millisecondTimeStamp];
```

Comment:
I took your suggestion of just dropping in the contacts offers at the bottom of the conversation _unless_ we're replacing an existing instance.